File: ppt_gen/preprocess.py
from __future__ import annotations

import os
import re
from pathlib import Path

from ppt_gen.cache import content_hash, is_cache_valid, write_cache_meta
from ppt_gen.mermaid import render_mermaid
from ppt_gen.paths import (
    BUILD_DIR,
    CACHE_DIR,
    DIAGRAMS_ASSETS_DIR,
    PLOTS_ASSETS_DIR,
    ensure_dirs,
)
from ppt_gen.registry import get_plot, load_data
from ppt_gen.save import save_figure
from ppt_gen.slide_context import SlideContext
from ppt_gen.table import get_table_data, render_table_markdown
from ppt_gen.theme.compile import compile_theme
from ppt_gen.theme.tokens import ThemeTokens, load_theme

DIRECTIVE_RE = re.compile(
    r"\{\{(plot|table|mermaid):([a-zA-Z0-9_/\-]+)(?:\s*\|\s*([^}]*))?\}\}"
)

OPTION_RE = re.compile(r"(\w+)\s*=\s*([^|]+)")


def _parse_options(options_str: str | None) -> dict[str, str]:
    if not options_str:
        return {}
    return {m.group(1).strip(): m.group(2).strip() for m in OPTION_RE.finditer(options_str)}


def _inject_branding_frontmatter(text: str, tokens: ThemeTokens) -> str:
    branding = tokens.raw.get("branding", {})
    footer = branding.get("footer")
    if not footer or not text.startswith("---"):
        return text

    end = text.find("---", 3)
    if end == -1:
        return text

    front = text[3:end]
    body = text[end + 3 :]
    lines = front.splitlines()
    has_footer = any(line.strip().startswith("footer:") for line in lines)
    if has_footer:
        return text

    insert_at = len(lines)
    for idx, line in enumerate(lines):
        if line.strip().startswith("paginate:"):
            insert_at = idx + 1
            break

    lines.insert(insert_at, f'footer: "{footer}"')
    return f"---\n" + "\n".join(lines) + f"\n---{body}"


def _parse_theme_from_deck(text: str) -> str:
    if not text.startswith("---"):
        return "scientific"
    end = text.find("---", 3)
    if end == -1:
        return "scientific"
    front = text[3:end]
    for line in front.splitlines():
        if line.strip().startswith("theme:"):
            return line.split(":", 1)[1].strip()
    return "scientific"


def _ensure_theme_compiled(theme_name: str) -> ThemeTokens:
    theme_path = Path(__file__).resolve().parent.parent / "themes" / theme_name
    tokens_file = theme_path / "tokens.yaml"
    css_file = theme_path / "theme.css"
    if not css_file.exists() or (
        tokens_file.exists() and tokens_file.stat().st_mtime > css_file.stat().st_mtime
    ):
        compile_theme(theme_name)
    return load_theme(theme_name)


def _render_plot_directive(
    name: str,
    options: dict[str, str],
    tokens: ThemeTokens,
    data: dict,
) -> str:
    slot = options.get("slot", "full")
    ctx = SlideContext(tokens=tokens, slot_name=slot, data=data)
    plot_fn = get_plot(name)
    source_path = Path(__file__).resolve().parent.parent / "plots" / f"{name.replace('/', '_')}.py"
    if not source_path.exists():
        # try nested name
        parts = name.split("/")
        source_path = Path(__file__).resolve().parent.parent / "plots" / f"{parts[-1]}.py"

    tokens_hash = content_hash(tokens.source_path.read_text(), tokens.name)
    data_hash = content_hash({k: str(v.shape) if hasattr(v, "shape") else str(v) for k, v in data.items()})
    plot_hash = content_hash(
        name,
        slot,
        tokens_hash,
        data_hash,
        source_path.read_text() if source_path.exists() else name,
    )
    out_path = PLOTS_ASSETS_DIR / f"{name.replace('/', '_')}.png"
    if not is_cache_valid(CACHE_DIR, f"plot-{name}", expected_hash=plot_hash, output_path=out_path):
        fig = plot_fn(ctx)
        save_figure(fig, ctx, out_path)
        write_cache_meta(CACHE_DIR, f"plot-{name}", hash_value=plot_hash, output=out_path)

    width_pct = tokens.figure_width_pct
    rel = _asset_href(out_path)
    return f"![width:{width_pct}%]({rel})"


def _asset_href(path: Path) -> str:
    return Path(os.path.relpath(path.resolve(), BUILD_DIR.resolve())).as_posix()


def _render_table_directive(
    name: str,
    options: dict[str, str],
    tokens: ThemeTokens,
    data: dict,
) -> str:
    max_rows = int(options["max_rows"]) if "max_rows" in options else None
    df = get_table_data(name, data)
    return render_table_markdown(df, tokens, max_rows=max_rows)


def _render_mermaid_directive(name: str, tokens: ThemeTokens) -> str:
    out_path = DIAGRAMS_ASSETS_DIR / f"{name}.svg"
    mermaid_src = Path(__file__).resolve().parent.parent / "mermaid" / f"{name}.mmd"
    tokens_hash = content_hash(tokens.source_path.read_text(), tokens.name)
    src_hash = content_hash(mermaid_src.read_text() if mermaid_src.exists() else name)
    diagram_hash = content_hash(name, tokens_hash, src_hash)

    if not is_cache_valid(
        CACHE_DIR, f"mermaid-{name}", expected_hash=diagram_hash, output_path=out_path
    ):
        render_mermaid(name, tokens, out_path)
        write_cache_meta(CACHE_DIR, f"mermaid-{name}", hash_value=diagram_hash, output=out_path)

    width_pct = tokens.figure_width_pct
    rel = _asset_href(out_path)
    return f"![width:{width_pct}%]({rel})"


def preprocess_deck(deck_path: Path, *, theme_name: str | None = None) -> Path:
    ensure_dirs()
    text = deck_path.read_text()
    theme = theme_name or _parse_theme_from_deck(text)
    tokens = _ensure_theme_compiled(theme)
    data = load_data()

    def replacer(match: re.Match[str]) -> str:
        kind, name, opts_raw = match.group(1), match.group(2), match.group(3)
        options = _parse_options(opts_raw)
        if kind == "plot":
            return _render_plot_directive(name, options, tokens, data)
        if kind == "table":
            return _render_table_directive(name, options, tokens, data)
        if kind == "mermaid":
            return _render_mermaid_directive(name, tokens)
        return match.group(0)

    processed = DIRECTIVE_RE.sub(replacer, text)
    processed = _inject_branding_frontmatter(processed, tokens)
    BUILD_DIR.mkdir(parents=True, exist_ok=True)
    out_path = BUILD_DIR / deck_path.name
    out_path.write_text(processed)
    return out_path

File: lib/preprocess.mjs
import { existsSync, statSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { relative, join, resolve } from "node:path";

import { contentHash, isCacheValid, writeCacheMeta } from "./cache.mjs";
import { SlideContext } from "./charts/context.mjs";
import { renderCsvPlot } from "./charts/from-csv.mjs";
import { loadData } from "./charts/registry.mjs";
import { savePngBuffer } from "./charts/helpers.mjs";
import { renderMermaid } from "./mermaid.mjs";
import {
  BUILD_DIR,
  CACHE_DIR,
  DIAGRAMS_ASSETS_DIR,
  MERMAID_DIR,
  PLOTS_ASSETS_DIR,
  ROOT,
  ensureDirs,
  themeDir,
} from "./paths.mjs";
import { getTableData, renderTableMarkdown } from "./table.mjs";
import { compileTheme } from "./theme/compile.mjs";
import { loadTheme } from "./theme/tokens.mjs";

const DIRECTIVE_RE =
  /\{\{(plot|table|mermaid):([a-zA-Z0-9_/\-]+)(?:\s*\|\s*([^}]*))?\}\}/g;
const OPTION_RE = /(\w+)\s*=\s*([^|]+)/g;
const IMG_RE = /!\[([^\]]*)\]\(([^)]+)\)/g;

function parseOptions(optionsStr) {
  if (!optionsStr) return {};
  const out = {};
  for (const m of optionsStr.matchAll(OPTION_RE)) {
    out[m[1].trim()] = m[2].trim();
  }
  return out;
}

function injectBrandingFrontmatter(text, tokens) {
  const branding = tokens.raw.branding ?? {};
  const footer = branding.footer;
  if (!footer || !text.startsWith("---")) return text;

  const end = text.indexOf("---", 3);
  if (end === -1) return text;

  const front = text.slice(3, end);
  const body = text.slice(end + 3);
  const lines = front.split(/\r?\n/);
  if (lines.some((line) => line.trim().startsWith("footer:"))) return text;

  let insertAt = lines.length;
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].trim().startsWith("paginate:")) {
      insertAt = i + 1;
      break;
    }
  }
  lines.splice(insertAt, 0, `footer: "${footer}"`);
  return `---\n${lines.join("\n")}\n---${body}`;
}

function parseThemeFromDeck(text) {
  if (!text.startsWith("---")) return "scientific";
  const end = text.indexOf("---", 3);
  if (end === -1) return "scientific";
  const front = text.slice(3, end);
  for (const line of front.split(/\r?\n/)) {
    if (line.trim().startsWith("theme:")) {
      return line.split(":").slice(1).join(":").trim();
    }
  }
  return "scientific";
}

async function ensureThemeCompiled(themeName) {
  const tokensFile = join(themeDir(themeName), "tokens.yaml");
  const cssFile = join(themeDir(themeName), "theme.css");
  if (
    !existsSync(cssFile) ||
    (existsSync(tokensFile) &&
      statSync(tokensFile).mtimeMs > statSync(cssFile).mtimeMs)
  ) {
    await compileTheme(themeName);
  }
  return loadTheme(themeName);
}

function assetHref(path, { absolute = false } = {}) {
  if (absolute) return resolve(path);
  return relative(BUILD_DIR, path).split("\\").join("/");
}

async function renderPlotDirective(name, options, tokens, data, hrefOpts) {
  if (!(name in data)) {
    throw new Error(
      `Unknown plot data '${name}'. Add data/${name}.csv (available: ${Object.keys(data).sort().join(", ")})`,
    );
  }
  const rows = data[name];
  const slot = options.slot ?? "full";
  const ctx = new SlideContext({ tokens, slotName: slot, data });

  const tokensText = await readFile(tokens.sourcePath, "utf8");
  const tokensHash = contentHash(tokensText, tokens.name);
  const dataHash = contentHash(rows);
  const optsHash = contentHash(options);
  const plotHash = contentHash(name, slot, tokensHash, dataHash, optsHash);

  const outStem = `${name.replace(/\//g, "_")}-${plotHash.slice(0, 10)}`;
  const outPath = join(PLOTS_ASSETS_DIR, `${outStem}.png`);
  if (
    !(await isCacheValid(CACHE_DIR, `plot-${outStem}`, {
      expectedHash: plotHash,
      outputPath: outPath,
    }))
  ) {
    const buffer = await renderCsvPlot(name, rows, options, ctx);
    await savePngBuffer(buffer, outPath);
    await writeCacheMeta(CACHE_DIR, `plot-${outStem}`, {
      hashValue: plotHash,
      output: outPath,
    });
  }

  return `![width:${tokens.figureWidthPct}%](${assetHref(outPath, hrefOpts)})`;
}

function renderTableDirective(name, options, tokens, data) {
  const maxRows =
    "max_rows" in options ? Number.parseInt(options.max_rows, 10) : null;
  const rows = getTableData(name, data);
  return renderTableMarkdown(rows, tokens, { maxRows });
}

async function renderMermaidDirective(name, tokens, hrefOpts) {
  const ext = hrefOpts.mermaidExt ?? "svg";
  const outPath = join(DIAGRAMS_ASSETS_DIR, `${name}.${ext}`);
  const mermaidSrc = join(MERMAID_DIR, `${name}.mmd`);
  const tokensText = await readFile(tokens.sourcePath, "utf8");
  const tokensHash = contentHash(tokensText, tokens.name);
  const srcHash = contentHash(
    existsSync(mermaidSrc) ? await readFile(mermaidSrc, "utf8") : name,
  );
  const diagramHash = contentHash(name, tokensHash, srcHash, ext);

  if (
    !(await isCacheValid(CACHE_DIR, `mermaid-${name}-${ext}`, {
      expectedHash: diagramHash,
      outputPath: outPath,
    }))
  ) {
    await renderMermaid(name, tokens, outPath);
    await writeCacheMeta(CACHE_DIR, `mermaid-${name}-${ext}`, {
      hashValue: diagramHash,
      output: outPath,
    });
  }

  return `![width:${tokens.figureWidthPct}%](${assetHref(outPath, hrefOpts)})`;
}

/**
 * Expand {{plot}} / {{table}} / {{mermaid}} in markdown text.
 * @param {object} [opts]
 * @param {boolean} [opts.absolute] — absolute image paths (for pptxgenjs)
 * @param {'svg'|'png'} [opts.mermaidExt]
 */
export async function expandDirectives(
  text,
  tokens,
  data,
  { absolute = false, mermaidExt = "svg" } = {},
) {
  const hrefOpts = { absolute, mermaidExt };
  const parts = [];
  let last = 0;
  for (const match of text.matchAll(DIRECTIVE_RE)) {
    parts.push(text.slice(last, match.index));
    const kind = match[1];
    const name = match[2];
    const options = parseOptions(match[3]);
    if (kind === "plot") {
      parts.push(await renderPlotDirective(name, options, tokens, data, hrefOpts));
    } else if (kind === "table") {
      parts.push(renderTableDirective(name, options, tokens, data));
    } else if (kind === "mermaid") {
      parts.push(await renderMermaidDirective(name, tokens, hrefOpts));
    } else {
      parts.push(match[0]);
    }
    last = match.index + match[0].length;
  }
  parts.push(text.slice(last));
  return parts.join("");
}

/**
 * Parse expanded markdown into contentSlide data (images, tables, bullets).
 */
export function parseContentFromExpandedMarkdown(markdown) {
  // Local heading extract (subtitle must not swallow directive/image lines)
  const lines = markdown.split(/\r?\n/);
  let title = null;
  let subtitle = null;
  let i = 0;
  while (i < lines.length && !lines[i].trim()) i++;
  const h = lines[i]?.match(/^#{1,2}\s+(.+)$/);
  if (h) {
    title = h[1].trim();
    i++;
    while (i < lines.length && !lines[i].trim()) i++;
    const line = lines[i];
    if (
      line &&
      !line.startsWith("#") &&
      !line.startsWith(":::") &&
      !line.startsWith("{{") &&
      !line.startsWith("!") &&
      !line.startsWith("|") &&
      !line.startsWith("<") &&
      !/^[-*]\s/.test(line)
    ) {
      subtitle = line.trim();
      i++;
    }
  }
  let working = lines.slice(i).join("\n").trim();
  const images = [];

  working = working.replace(IMG_RE, (_m, alt, src) => {
    let path = src.trim();
    if (!path.startsWith("/") && !/^[A-Za-z]:/.test(path)) {
      const fromBuild = resolve(BUILD_DIR, path);
      const fromRoot = resolve(ROOT, path);
      if (existsSync(fromBuild)) path = fromBuild;
      else if (existsSync(fromRoot)) path = fromRoot;
      else path = fromBuild;
    }
    const widthMatch = String(alt).match(/width:(\d+)%/);
    const widthPct = widthMatch ? Number(widthMatch[1]) : 90;
    const w = Math.round(1156 * (widthPct / 100));
    images.push({ path, w, h: 400 });
    return "";
  });

  working = working.replace(/<style[\s\S]*?<\/style>/gi, "").trim();

  let table = null;
  const tableLines = working.split(/\r?\n/);
  const tableStart = tableLines.findIndex((l) => /^\|.+\|$/.test(l.trim()));
  if (tableStart >= 0) {
    const splitRow = (row) =>
      row
        .split("|")
        .slice(1, -1)
        .map((c) => c.trim());
    const headers = splitRow(tableLines[tableStart]);
    let ti = tableStart + 1;
    if (tableLines[ti] && /^\|[\s:|-]+\|$/.test(tableLines[ti].trim())) ti += 1;
    const rows = [];
    while (ti < tableLines.length && /^\|.+\|$/.test(tableLines[ti].trim())) {
      rows.push(splitRow(tableLines[ti]));
      ti += 1;
    }
    table = { headers, rows };
    const before = tableLines.slice(0, tableStart).join("\n");
    const after = tableLines.slice(ti).join("\n");
    working = `${before}\n${after}`.trim();
  }

  const bullets = [];
  const bodyParts = [];
  for (const line of working.split(/\r?\n/)) {
    const b = line.match(/^[-*]\s+(.+)$/);
    if (b) {
      bullets.push(b[1].replace(/\*\*(.+?)\*\*/g, "$1"));
    } else if (line.trim() && !line.startsWith("#")) {
      bodyParts.push(line.replace(/\*\*(.+?)\*\*/g, "$1").trim());
    }
  }

  if (images.length === 1 && !table && bullets.length === 0) {
    images[0].h = 440;
  } else if (images.length) {
    for (const img of images) img.h = 360;
  }

  return {
    title,
    subtitle,
    images,
    table,
    bullets,
    body: bodyParts.filter(Boolean).join("\n") || null,
  };
}

export async function preprocessDeck(deckPath, { themeName = null } = {}) {
  await ensureDirs();
  let text = await readFile(deckPath, "utf8");

  // Shared IR: expand ::: construct blocks to Marp-readable fallbacks first.
  const { parseDeckMarkdown } = await import("./ir/parse.mjs");
  const { irToMarpMarkdown } = await import("./ir/to-marp.mjs");
  const ir = parseDeckMarkdown(text);
  if (ir.hasConstructs) {
    text = irToMarpMarkdown(ir);
  }

  const theme = themeName || parseThemeFromDeck(text) || ir.meta.theme || "scientific";
  const tokens = await ensureThemeCompiled(theme);
  const data = await loadData();

  let processed = await expandDirectives(text, tokens, data, {
    absolute: false,
    mermaidExt: "svg",
  });
  processed = injectBrandingFrontmatter(processed, tokens);
  await mkdir(BUILD_DIR, { recursive: true });
  const outPath = join(BUILD_DIR, deckPath.split(/[/\\]/).pop());
  await writeFile(outPath, processed);
  return outPath;
}

export { ensureThemeCompiled, parseThemeFromDeck };
export { loadData } from "./charts/registry.mjs";
